File: php/tests/Unit/Advertisement/Domain/ValueObject/DescriptionTest.php
<?php
declare(strict_types=1);

namespace Tests\Demo\App\Unit\Advertisement\Domain\ValueObject;

use Demo\App\Advertisement\Domain\Exceptions\DescriptionEmptyException;
use Demo\App\Advertisement\Domain\Exceptions\DescriptionTooLongException;
use Demo\App\Advertisement\Domain\Model\ValueObject\Description;
use PHPUnit\Framework\TestCase;
use ReflectionClass;

class DescriptionTest extends TestCase
{
    private const string DESCRIPTION = 'description';

    public function testShouldNotBeInstantiatedWithTheConstructor(): void
    {
        $class = new ReflectionClass(Description::class);
        $constructor = $class->getConstructor();

        self::assertTrue($constructor->isPrivate(), 'Constructor is not private');
    }

    public function testShouldCreateADescription()
    {
        $result = Description::build(self::DESCRIPTION);

        $description = $result->unwrap();
        self::assertInstanceOf(Description::class, $description);
        self::assertEquals(self::DESCRIPTION, $description->value());
    }

    public function testShouldReturnErrorResultWhenDescriptionHasMoreThan200Characters()
    {
        $randomString = str_repeat('a', 201);
        $result = Description::build($randomString);

        self::assertFalse($result->isSuccess());
        self::assertInstanceOf(DescriptionTooLongException::class, $result->exception());
        self::assertEquals('Description has more than 200 characters: Has 201 characters', $result->exception()->getMessage());
    }

    public function testShouldReturnErrorResultWhenDescriptionIsEmpty()
    {
        $result = Description::build('');

        self::assertFalse($result->isSuccess());
        self::assertInstanceOf(DescriptionEmptyException::class, $result->exception());
        self::assertEquals('Description empty', $result->exception()->getMessage());
    }
}

File: kotlin/src/test/kotlin/unit/advertisement/domain/model/value_object/DescriptionTest.kt
package unit.advertisement.domain.model.value_object

import advertisement.domain.errors.DescriptionEmptyError
import advertisement.domain.errors.DescriptionTooLongError
import advertisement.domain.model.value_object.Description
import arrow.core.*
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.lang.reflect.Modifier


class DescriptionTest
{
    companion object {
        private const val VALID_DESCRIPTION = "Description test"
        private const val EMPTY_DESCRIPTION = ""
        private val LONG_DESCRIPTION = "a".repeat(201)
    }

    @Test
    fun testShouldNotBeInstantiatedWithTheConstructor() {
        Assertions.assertThrows(NoSuchMethodException::class.java) {
            Assertions.assertTrue(Modifier.isPrivate(Description::class.java.getDeclaredConstructor().modifiers))
        }
    }

    @Test
    fun testShouldCreateADescription() {
        val result = Description.build(VALID_DESCRIPTION)

        result.fold(
            { error -> Assertions.fail("Expected a valid description, but got error: $error") },
            { description -> Assertions.assertEquals(VALID_DESCRIPTION, description.value()) }
        )
    }

    @Test
    fun testShouldReturnErrorForEmptyDescription() {
        val result = Description.build(EMPTY_DESCRIPTION)

        Assertions.assertTrue(result is Either.Left)
        result.fold(
            { error ->
                Assertions.assertTrue(error is DescriptionEmptyError)
                Assertions.assertEquals("Description empty", error.message)
            },
            { description -> Assertions.fail("Expected an error, but got a valid description: ${description.value()}") }
        )
    }

    @Test
    fun testShouldReturnErrorForTooLongDescription() {
        val result = Description.build(LONG_DESCRIPTION)

        Assertions.assertTrue(result is Either.Left)
        result.fold(
            { error ->
                Assertions.assertTrue(error is DescriptionTooLongError)
                Assertions.assertEquals("Description has more than 200 characters: Has ${LONG_DESCRIPTION.length}", error.message)
            },
            { description -> Assertions.fail("Expected an error, but got a valid description: ${description.value()}") }
        )
    }
}
